test/dynamo/test_export.py

```diff
@@ -3124,6 +3110,136 @@ def forward(self, x):
     return pytree.tree_unflatten([slice_tensor, slice_tensor_3, slice_tensor_6, slice_tensor_9], self._out_spec)""",
         )
 
+    def test_cond_op_param_buffer_lifted(self):
+        class A(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.register_buffer("buffer1", torch.zeros(6, 4))
+
+            def forward(self):
+                return self.buffer1.sum()
+
+        class B(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.register_buffer("buffer2", torch.ones(6, 4))
+
+            def forward(self):
+                return self.buffer2.sum()
+
+        class M(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.a = A()
+                self.b = B()
+
+            def forward(self, x):
+                def true_fn(x):
+                    return x.cos() + self.a()
+
+                def false_fn(x):
+                    return x.sin() + self.b()
+
+                return (cond(x.shape[0] > 4, true_fn, false_fn, [x]),)
+
+        gm, _ = torch._dynamo.export(M(), torch.ones(6, 4), aten_graph=False)
+        self.assertEqual(gm(torch.ones(6, 4)), M()(torch.ones(6, 4)))
+        self.assertEqual(gm(torch.ones(3, 4)), M()(torch.ones(3, 4)))
+
```

Comment:
All of the tests are for export. We should add tests for training as well under test/dynamo/test_higher_order_ops.py.

I'm not really sure what the best way to prevent overlap these test_export and test_higher_order_ops is. Morally tests for training should be a superset of the tests for export, so I would actually toss all of the heavy testing into test_higher_order_ops, but it sounds like export goes through a slightly different mechanism?

Things we should test:
- map (we can go light on these tests, because we have a straightforward application of speculate_subgraph)
   - map with captured value
   - map with side effects in body
   - map with graph break in body
- cond (there are a bunch of interesting tests here)
    - graph break in true or false branch
    - true branch has free variables
    - false branch has free variables
    - true and false branch both have free variables
    - true and false branch capture free variables (with some overlap)
    - true branch has side effects
    - false branch has side effects
    - only the true branch or the false branch has a guard (I'm not really sure how to write this test, but we want to make sure that this actually invokes a recompile or turns into a graph break if it's not implemented)

Reply:
Yep I think I covered majority of the cases.